src/other/index.jl

```diff
@@ -108,19 +108,49 @@ function Base.insert!(x::Index, idx::Integer, nm::Symbol)
     x
 end
 
-Base.getindex(x::Index, idx::Symbol) = x.lookup[idx]
-Base.getindex(x::AbstractIndex, idx::Real) = Int(idx)
-Base.getindex(x::AbstractIndex, idx::AbstractVector{Union{Bool, Missing}}) =
-    getindex(x, collect(Missings.replace(idx, false)))
+Base.getindex(x::AbstractIndex, idx::Symbol) = x.lookup[idx]
+Base.getindex(x::AbstractIndex, idx::AbstractVector{Symbol}) = [x.lookup[i] for i in idx]
+
+function Base.getindex(x::AbstractIndex, idx::Real)
+    idx isa Bool && Base.depwarn("Indexing with Bool values is deprecated", :getindex)
+    idx isa Integer || Base.depwarn("Indexing with values that are not Integer is deprecated", :getindex)
+    Int(idx)
+end
+
+function Base.getindex(x::AbstractIndex, idx::AbstractVector{<:Real})
+    any(v -> v isa Bool, idx) && Base.depwarn("Indexing with Bool values is deprecated")
+    all(v -> v isa Integer, idx) || Base.depwarn("Indexing with values that are not Integer is deprecated", :getindex)
+    Vector{Int}(idx)
+end
+
 function Base.getindex(x::AbstractIndex, idx::AbstractVector{Bool})
     length(x) == length(idx) || throw(BoundsError(x, idx))
     find(idx)
 end
-Base.getindex(x::AbstractIndex, idx::AbstractVector{T}) where {T >: Missing} =
-    getindex(x, collect(skipmissing(idx)))
-Base.getindex(x::AbstractIndex, idx::AbstractRange) = [idx;]
-Base.getindex(x::AbstractIndex, idx::AbstractVector{T}) where {T <: Real} = convert(Vector{Int}, idx)
-Base.getindex(x::AbstractIndex, idx::AbstractVector{Symbol}) = [x.lookup[i] for i in idx]
+
+Base.getindex(x::AbstractIndex, idx::AbstractRange) = getindex(x, collect(idx))
```

Comment:
The old method was more efficient, why call `getindex`?

Reply:
because `[idx;]` may produce something else that `Vector{Int}` and the contract of `getindex` is that it returns either `Int` or `Vector{Int}`.

Reply:
But not why just `convert(Vector{Int}, idx)`? More radically, why not change the contract and decide it can return any `AbstractVector{Int}`? Doesn't sound very useful to convert to a vector.

Reply:
It is not a `Vector` part that is a problem, but `Int` part.

If I simply convert then `1.0:3.0` or `true:true` will silently pass though as valid index ranges and the should not (at least based on what is consistent with Base). Currently it is converted to `Vector` so the second round of dispatching does it work to check if indexing is valid.

Exactly because of such problems  I think that in general it is better to delegate handling of indexing to Base except what we have special in `DaraFrames` (i.e. indexing by `Symbol` or `AbstractVector{Symbol}`).

Reply:
Actually, wouldn't it be possible to have a method for `::AbstractRange{<:Integer}` which will just return its input, and one for `::AbstractRange` which will call `collect`?

Reply:
`stack` in reshape.jl expects `Vector{Int}` explicitly now (and relies on the conversion made here). I can change it.
Also we would have to handle a special case of `:AbstractRange{Bool}` separately but it should not be a problem.

Should I go ahead with those changes?

Reply:
In general better require just `::AbstractVector{Integer}`, yes.

Reply:
refactored. I continue to `collect` general ranges of `AbstractRange{<:Integer}` other than `AbstractRange{Int}` to cast them to `Int` (so that we have always underlying `Int` returned), but it can be changed.